src/common.jl

```diff
@@ -48,3 +48,93 @@ parameterless_type(x) = parameterless_type(typeof(x))
 parameterless_type(x::Type) = __parameterless_type(x)
 
 @non_differentiable adapt_randn(::Any...)
+
+
+"""
+    Closure{F,G}
+
+A callable of the form `(x, args...) -> F(G(args...), x)`.
+
+# Examples
+
+This is particularly useful when one wants to avoid broadcasting over constructors
+which can sometimes cause issues with type-inference, in particular when combined
+with reverse-mode AD frameworks.
+
+```juliarepl
+julia> using DistributionsAD, Distributions, ReverseDiff, BenchmarkTools
+
+julia> const data = randn(1000);
+
+julia> x = randn(length(data));
+
+julia> f(x) = sum(logpdf.(Normal.(x), data))
+f (generic function with 2 methods)
+
+julia> @btime ReverseDiff.gradient(\$f, \$x);
+  848.759 μs (14605 allocations: 521.84 KiB)
+
+julia> # Much faster with ReverseDiff.jl.
+       g(x) = sum(DistributionsAD.Closure(logpd, Normal).(data, x))
+g (generic function with 1 method)
+
+julia> @btime ReverseDiff.gradient(\$g, \$x);
+  17.460 μs (17 allocations: 71.52 KiB)
+```
+
+See https://github.com/TuringLang/Turing.jl/issues/1934 more further discussion.
+"""
+struct Closure{F,G} end
+
+Closure(::F, ::G) where {F,G} = Closure{F,G}()
+Closure(::F, ::Type{G}) where {F,G} = Closure{F,G}()
+Closure(::Type{F}, ::G) where {F,G} = Closure{F,G}()
+Closure(::Type{F}, ::Type{G}) where {F,G} = Closure{F,G}()
+
+"""
+    is_diff_safe(f)
+
+Return `true` if it's safe to ignore gradients wrt. `f` when computing `f`.
+
+Useful for checking it's okay to take faster paths in pullbacks for certain AD backends.
+
+# Examples
+
+```jldoctest
+julia> using Distributions
+
+julia> using DistributionsAD: is_diff_safe, Closure
+
+julia> is_diff_safe(typeof(logpdf))
+true
+
+julia> is_diff_safe(typeof(x -> 2x))
+true
+
+julia> # But it fails if we make a closure over a variable, which we might want to compute
+       # the gradient with respect to.
+       makef(x) = y -> x + y
+makef (generic function with 1 method)
+
+julia> is_diff_safe(typeof(makef([1.0])))
+false
+
+julia> # Also works on `Closure`s from `DistributionsAD`.
+       is_diff_safe(typeof(Closure(logpdf, Normal)))
+true
+
+julia> is_diff_safe(typeof(Closure(logpdf, makef([1.0]))))
+false
+"""
+@inline is_diff_safe(_) = false
```

Comment:
It seems this serves the same purpose as `ReverseDiff.mayhavetracked`: https://github.com/JuliaDiff/ReverseDiff.jl/blob/d522508aa6fea16e9716607cdd27d63453bb61e6/src/derivatives/broadcast.jl#L29

Reply:
Yeah basically, though I was looking at Zygote's `_dual_purefun` because ReverseDiff already hits the `broadcast_forward`, even without the custom adjoint I defined (ReverseDiff also doesn't support calling back into AD).